### PR TITLE
Add new GitHub Actions workflow to test if old releases install without error

### LINKED ISSUE #4696: Pip update (`24.1`) incompatible with `dipy==1.5.0`, causing errors for old SCT releases
Forum post: https://forum.spinalcordmri.org/t/problem-installing-sct-v6-1/1337

Traceback:

```python
WARNING: Ignoring version 1.5.0 of dipy since it has invalid metadata:
# Requested dipy==1.5.0 from https://files.pythonhosted.org/packages/4d/ff/933d40f73187c3e1996e833d8fe20c0863130e8d881e09bba14449a42e85/dipy-1.5.0-cp39-cp39-manylinux_2_12_x86_64.manylinux2010_x86_64.whl (from -r requirements-freeze.txt (line 17)) 
# has invalid metadata: Expected matching RIGHT_PARENTHESIS for LEFT_PARENTHESIS, after version specifier
    fury (>=0.8.0scikit-learn) ; extra == 'all'
         ~~~~~~~~^
Please use pip<24.1 if you need to use this version.
ERROR: Ignored the following versions that require a different python version: 1.3.1 Requires-Python >=3.10
ERROR: Could not find a version that satisfies the requirement dipy==1.5.0 (from versions: 0.5.0, 0.6.0, 0.7.0, 0.7.1, 0.8.0, 0.9.1, 0.9.2, 0.10.1, 0.11.0, 0.12.0, 0.13.0, 0.14.0, 0.15.0, 0.16.0, 1.0.0, 1.1.0, 1.2.0, 1.3.0, 1.4.0, 1.4.1, 1.5.0, 1.6.0, 1.7.0, 1.8.0, 1.9.0)
ERROR: No matching distribution found for dipy==1.5.0
```

The underlying issue here is that we freeze specific versions of our dependencies, but we _don't_ freeze `pip` itself, meaning that future `pip` upgrades can retroactively break old frozen versions of packages. 

So, ideally, in addition to limiting `pip` here, we could also consider freezing our `conda` dependencies as part of the stable release procedure. :thinking:

### PR DESCRIPTION
**Pull request by joshuacwnewton**
## Description

This PR is a helper PR, so that I can see where the issues lie with our old releases. We can then take these fixes (or alternative fixes depending on the review) and backport them to the corresponding tags.

This is an expensive PR (huge matrix of OSs and versions), so ideally my hope is that we run this as a cron job maybe... weekly? So that we can feel safe knowing that older versions still work, without using up a ton of runners on every PR.

We can also use this workflow to test backported changes before we merge them. (e.g. if we make a branch off of `5.4` called e.g. `jn/5.4-backport-fix`, we can also temporarily update this workflow and add `jn/5.4-backport-fix` to easily install and test that version, without having to fuss around with installing the older versions locally.

## Linked issues

Partially addresses #4696.
Partially addresses https://forum.spinalcordmri.org/t/problem-installing-sct-v5-8-on-linux-mint-21-1/1367.

### CODE REVIEW

File: .github/workflows/test-past-releases.yml
Comment (joshuacwnewton) [inline .github/workflows/test-past-releases.yml:1]: Huzzah! After many tweaks and iterations, I've finally added enough fixes to get the workflow running on all versions of macOS/Windows/Ubuntu (sans some pending macOS runners, since we've hit our limit).

I'd like to add testing for WSL, too, since it came up in a recent case where a reviewer tried to validate the code of one of our lab's publications and ran into an error on installing SCT v6.3 on WSL.

But, for now, I think I'll cancel the in-progress workflow (to free up macOS runners), clean up the PR with some rebasing, and mark it as ready for review.

File: .github/workflows/test-past-releases.yml [hunk 1]
```diff
@@ -0,0 +1,168 @@
+name: Test past releases
+
+on:
+  schedule:
+    # https://docs.github.com/en/free-pro-team@latest/actions/reference/workflow-syntax-for-github-actions#onschedule
+    # > Scheduled workflows run on the latest commit on the default or base branch
+    # i.e. this can only run on master
+    - cron:  '0 11 * * *'
```
Comment (mguaypaq) [inline .github/workflows/test-past-releases.yml:8]: I agree with your suggestion of running this weekly, but I think the current cron line says "11am UTC every day" (that is, 6am EST or 7am EDT). I believe this would limit it to weekly on Sundays (presumably a not-busy day for us):
```suggestion
    - cron:  '0 11 * * 0'
```

**Review by mguaypaq — APPROVED** [review]
Wow, thanks for patiently handling all the corner cases of our past installation methods and various bugs. This is great!

File: .github/workflows/test-past-releases.yml [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+        run: |
+          # Add fix for clash between pip 24.1 and dipy 1.5.0
+          perl -i -pe 's/pip!=21.2.\*/pip!=21.2.\*,<24.1" "setuptools[core]/g;' install_sct
+          ./install_sct -y
+
+      - name: Windows hotfix for >=5.9 (separate bc bash)
+        if: runner.os == 'Windows' && matrix.sct_version != '5.7' && matrix.sct_version != '5.8'
+        shell: bash
+        # Versions >=5.9 use the new syntax ("-U pip!=21.2) and also need the '^`' escape for the '!' character.
+        run: perl -i -pe 's/pip\^!=21.2.\*/pip^!=21.2.\*,<24.1" "setuptools[core]/g;' install_sct.bat
```
Comment (joshuacwnewton) [inline .github/workflows/test-past-releases.yml:94]: @mguaypaq @SomeoneInParticular: Right now, the `pip`/`setuptools` fixes are only placing a hard upper limit on `pip`, without placing limits on `packaging`, `setuptools`, etc.

I am wondering if we should instead take a look at which packages get installed for each version of SCT being tested, and freeze the `conda` packages exactly as they are in the "create conda environment" step. (For context, we freeze our `pip` packages (numpy, torch, etc.) but we don't freeze the conda toolchain that installs the aforementioned pip packages.)

For example, for macOS + SCT v5.7, we see:

```
  ca-certificates    pkgs/main/osx-64::ca-certificates-2024.12.31-hecd8cb5_0 
  certifi            pkgs/main/osx-64::certifi-2022.12.7-py37hecd8cb5_0 
  libffi             pkgs/main/osx-64::libffi-3.4.4-hecd8cb5_1 
  ncurses            pkgs/main/osx-64::ncurses-6.4-hcec6c5f_0 
  openssl            pkgs/main/osx-64::openssl-1.1.1w-hca72f7f_0 
  pip                pkgs/main/osx-64::pip-22.3.1-py37hecd8cb5_0 
  python             pkgs/main/osx-64::python-3.7.16-h218abb5_0 
  readline           pkgs/main/osx-64::readline-8.2-hca72f7f_0 
  setuptools         pkgs/main/osx-64::setuptools-65.6.3-py37hecd8cb5_0 
  sqlite             pkgs/main/osx-64::sqlite-3.45.3-h6c40b1e_0 
  tk                 pkgs/main/osx-64::tk-8.6.14-h4d00af3_0 
  wheel              pkgs/main/osx-64::wheel-0.38.4-py37hecd8cb5_0 
  xz                 pkgs/main/osx-64::xz-5.4.6-h6c40b1e_1 
  zlib               pkgs/main/osx-64::zlib-1.2.13-h4b97444_1 
```

But then we update `pip` and `setuptools[core]`:

```
Successfully installed pip-24.0 setuptools-68.0.0
```

Would freezing (or alternatively, not updating pip/setuptools at all) be preferable here, to avoid any future issues when setting up old releases?

**Review by joshuacwnewton — COMMENTED** [review]
Before merging, I just wanted to double-check the solution we plan to backport to the older releases.

File: .github/workflows/test-past-releases.yml (hunk 2 above)
Comment (joshuacwnewton) [reply .github/workflows/test-past-releases.yml:90]: In today's meeting, we voted in favor of freezing the toolchain dependencies for safety/peace of mind reasons. I'll merge this PR and incorporate that decision into my backporting efforts.
